Add integer overflow helper functions

paulidale opened the pull request:
Add a header only sub-system to simplify the detection and processing of integer operations that risk overflow.

Currently functions are defined for the basic arithmetic operations `+`, `-`, `*`, `/` and `%`.  In each case the function returns the result of the arithmetic operation if such a result is not problematic.  When a result is problematic (overflow, underflow, division by zero), a value is returned and the error flag is set.  This permits a series of operations to be undertaken and the error flag checked at the end to determine if any failed.

The functions also guarantee that no operations undefined by the C standard occur during the calculation of the result and error status.  Specifically, signed integer operations that overflow are undefined by the standard.  There is a guarantee that no such operation will occur.

- [x] documentation is added or updated
- [x] tests are added or updated

paulidale commented:
Ping for reviews?

t8m commented on include/internal/safe_math.h:
```diff
@@ -0,0 +1,320 @@
+
+# include <openssl/e_os2.h>              /* For 'ossl_inline' */
+
+# ifndef OSSL_NO_OVERFLOW_CHECKING
```
Would this be first OSSL_NO macro? Perhaps keeping OPENSSL_NO_.... is more aligned with the rest.

paulidale replied:
The name isn't good, this is enabling the use of builtin overflow check functions.  It allows the C versions to be tested.  Alternatively, I could drop this and rely on the CIs to cover all paths (I'm not sure they do currently).

paulidale replied:
Changed this to `OPENSSL_NO_BUILTIN_OVERFLOW_CHECKING` which is what it actually does.

t8m reviewed (changes requested):
Just one nit. Should we merge this only once we want to start using this API or do we want to do that in advance?

paulidale commented:
I no strong feelings either way about merging this now or later.

I'd like to implement a fused multiply/divide operation since that's what I wanted initially.

paulidale commented:
Fused multiply/divide added.

paulidale commented:
I didn't notice the approval :(

I added some better failure return values, documented the muldiv function and added a run-checker test for the not compiler supported path.

kroeckx commented on crypto/stack/stack.c:
```diff
@@ -138,32 +141,33 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
-        current = current < limit ? current + current / 2 : max_nodes;
+        current = safe_muldiv_int(current, 8, 5, &err);
+        if (err)
+            return 0;
```
This can result in a current > max_nodes, while the old code returned max_nodes in that case.

paulidale replied:
Good point.

paulidale replied:
Fixed now.

kroeckx reviewed (commented):
I would like to see more code changes to use the APIs.

paulidale commented:
In time....

paulidale commented:
Merged, thanks for the feedback.
